common/web/gesture-recognizer/src/inputEventEngine.ts

```diff
@@ -18,20 +20,70 @@ namespace com.keyman.osk {
     abstract registerEventHandlers();
     abstract unregisterEventHandlers();
 
-    onInputStart(coord: InputEventCoordinate) {
-      this.emit(InputEventEngine.INPUT_UPDATE_EVENT_NAME, TrackedInputState.START, coord);
+    /**
+     * @param identifier The identifier number corresponding to the input sequence.
+     */
+    hasActiveTouchpoint(identifier: number) {
+      return this._activeTouchpoints.findIndex((point) => point.rawIdentifier == identifier) != -1;
+    }
+
+    private getTouchpointWithId(identifier: number) {
+      return this._activeTouchpoints.find((point) => point.rawIdentifier == identifier);
+    }
+
+    public dropTouchpointWithId(identifier: number) {
+      this._activeTouchpoints = this._activeTouchpoints.filter((point) => point.rawIdentifier != identifier);
+    }
+
+    protected buildSampleFor(clientX, clientY): InputSample {
+      const targetRect = this.config.targetRoot.getBoundingClientRect();
+      return {
+        clientX: clientX,
+        clientY: clientY,
+        targetX: clientX - targetRect.left,
+        targetY: clientY - targetRect.top,
+        t: performance.now()
+      };
     }
 
-    onInputMove(coord: InputEventCoordinate) {
-      this.emit(InputEventEngine.INPUT_UPDATE_EVENT_NAME, TrackedInputState.MOVE, coord);
+    onInputStart(identifier: number, sample: InputSample, target: EventTarget) {
+      const touchpoint = new TrackedPoint(identifier, target, this instanceof TouchEventEngine);
+      touchpoint.path.extend(sample);
+
+      this._activeTouchpoints.push(touchpoint);
+
+      // External objects may desire to directly terminate handling of
+      // input sequences under specific conditions.
+      touchpoint.path.on('invalidated', () => {
+        this.dropTouchpointWithId(identifier);
+      });
+
+      touchpoint.path.on('complete', () => {
+        this.dropTouchpointWithId(identifier);
+      });
+
+      this.emit('pointstart', touchpoint);
     }
 
-    onInputMoveCancel(coord: InputEventCoordinate) {
-      this.emit(InputEventEngine.INPUT_UPDATE_EVENT_NAME, TrackedInputState.CANCEL, coord);
+    onInputMove(identifier: number, sample: InputSample) {
+      this.getTouchpointWithId(identifier)?.path.extend(sample);
+    }
+
+    onInputMoveCancel(identifier: number, sample: InputSample) {
+      const touchpoint = this.getTouchpointWithId(identifier);
+
+      if(!touchpoint) {
+        return;
+      }
+
+      touchpoint.path.extend(sample);
+      touchpoint.path.terminate(true);
```

Comment:
```suggestion
      const touchpoint = this.getTouchpointWithId(identifier);
      touchpoint?.path.extend(sample);
      touchpoint?.path.terminate(true);
```

This better matches the style of `onInputMove`.